client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -389,12 +428,30 @@ where
 	let mut finalized_block_hashes = notification.tree_route.iter().cloned().collect::<Vec<_>>();
 	finalized_block_hashes.push(last_finalized);
 
-	let pruned_block_hashes: Vec<_> = notification.stale_heads.iter().cloned().collect();
+	// Report all pruned blocks from the notification that are not
+	// part of the `pruned_forks`.
+	let pruned_block_hashes: Vec<_> = {
+		let mut to_ignore = pruned_forks.write();
+
+		notification
+			.stale_heads
+			.iter()
+			.cloned()
```

Comment:
- Nit: Move `cloned()` after `filter()` to clone only what is necessary.
- SuperNit: Also the closure can be written as: `.filter(|hash| !to_ignore.remove(hash))`  (I know, the intent is more cryptic, but not that much 😃
- Also, as @skunert said, within the finality notification you find only the displaced leaves thus in this way you're not going to ignore the intermediate blocks that joins the displaced leaf down to the canonical chain.